cmake/introspection.cmake

```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+
+if(HAVE_BYTESWAP_H)
+  check_cxx_symbol_exists(bswap_16 "byteswap.h" HAVE_DECL_BSWAP_16)
+  check_cxx_symbol_exists(bswap_32 "byteswap.h" HAVE_DECL_BSWAP_32)
+  check_cxx_symbol_exists(bswap_64 "byteswap.h" HAVE_DECL_BSWAP_64)
+endif()
+
+if(HAVE_ENDIAN_H OR HAVE_SYS_ENDIAN_H)
+  if(HAVE_ENDIAN_H)
+    set(ENDIAN_HEADER "endian.h")
+  else()
+    set(ENDIAN_HEADER "sys/endian.h")
+  endif()
+  check_cxx_symbol_exists(be16toh ${ENDIAN_HEADER} HAVE_DECL_BE16TOH)
+  check_cxx_symbol_exists(be32toh ${ENDIAN_HEADER} HAVE_DECL_BE32TOH)
+  check_cxx_symbol_exists(be64toh ${ENDIAN_HEADER} HAVE_DECL_BE64TOH)
+  check_cxx_symbol_exists(htobe16 ${ENDIAN_HEADER} HAVE_DECL_HTOBE16)
+  check_cxx_symbol_exists(htobe32 ${ENDIAN_HEADER} HAVE_DECL_HTOBE32)
+  check_cxx_symbol_exists(htobe64 ${ENDIAN_HEADER} HAVE_DECL_HTOBE64)
+  check_cxx_symbol_exists(htole16 ${ENDIAN_HEADER} HAVE_DECL_HTOLE16)
+  check_cxx_symbol_exists(htole32 ${ENDIAN_HEADER} HAVE_DECL_HTOLE32)
+  check_cxx_symbol_exists(htole64 ${ENDIAN_HEADER} HAVE_DECL_HTOLE64)
+  check_cxx_symbol_exists(le16toh ${ENDIAN_HEADER} HAVE_DECL_LE16TOH)
+  check_cxx_symbol_exists(le32toh ${ENDIAN_HEADER} HAVE_DECL_LE32TOH)
+  check_cxx_symbol_exists(le64toh ${ENDIAN_HEADER} HAVE_DECL_LE64TOH)
+endif()
+
+CHECK_INCLUDE_FILE_CXX(unistd.h HAVE_UNISTD_H)
+if(HAVE_UNISTD_H)
+  check_cxx_symbol_exists(fdatasync "unistd.h" HAVE_FDATASYNC)
+  check_cxx_symbol_exists(fork "unistd.h" HAVE_DECL_FORK)
+  check_cxx_symbol_exists(pipe2 "unistd.h" HAVE_DECL_PIPE2)
+  check_cxx_symbol_exists(setsid "unistd.h" HAVE_DECL_SETSID)
+endif()
+
+CHECK_INCLUDE_FILE_CXX(sys/types.h HAVE_SYS_TYPES_H)
+CHECK_INCLUDE_FILE_CXX(ifaddrs.h HAVE_IFADDRS_H)
+if(HAVE_SYS_TYPES_H AND HAVE_IFADDRS_H)
+  check_cxx_symbol_exists(freeifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_FREEIFADDRS)
+  check_cxx_symbol_exists(getifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_GETIFADDRS)
+  # Illumos/SmartOS requires linking with -lsocket if
+  # using getifaddrs & freeifaddrs.
+  # See: https://github.com/bitcoin/bitcoin/pull/21486
+  if(HAVE_DECL_GETIFADDRS AND HAVE_DECL_FREEIFADDRS)
+    set(check_socket_source "
+      #include <sys/types.h>
+      #include <ifaddrs.h>
+
+      int main() {
+        struct ifaddrs* ifaddr;
+        getifaddrs(&ifaddr);
+        freeifaddrs(ifaddr);
+      }
+    ")
+    check_cxx_source_compiles("${check_socket_source}" IFADDR_WORKS_WITHOUT_LINK)
+    if(NOT IFADDR_WORKS_WITHOUT_LINK)
+      set(CMAKE_REQUIRED_LIBRARIES socket)
+      check_cxx_source_compiles("${check_socket_source}" IFADDR_NEEDS_LINK_TO_SOCKET)
+      set(CMAKE_REQUIRED_LIBRARIES)
+      if(IFADDR_NEEDS_LINK_TO_SOCKET)
+        link_libraries(socket)
+      else()
+        message(FATAL_ERROR "Cannot figure out how to use getifaddrs/freeifaddrs.")
+      endif()
+    endif()
+  endif()
+endif()
+
+# Check for gmtime_r(), fallback to gmtime_s() if that is unavailable.
+# Fail if neither are available.
+check_cxx_source_compiles("
```

Comment:
These are actually behavioral changes because CMake (surprisingly imo) invokes the linker for `check_cxx_source_compiles`.

We may want to investigate `set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)` at the top-level to disable that, then opting-in to linking as necessary (like for the socket check above).

Reply:
[Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436930235).